### PR TITLE
Added additional functional tests for TempData

### PR DESCRIPTION
**Pull request by ajaybhargavb**
Issue #2157

### CODE REVIEW

File: test/WebSites/TempDataWebSite/Controllers/HomeController.cs [hunk 1]
```diff
@@ -17,5 +18,30 @@ public IActionResult DisplayTempData(string value)
...
+        public IActionResult SetTempDataAndRedirectToPeek(string value)
+        {
+            TempData["key"] = value;
+            return RedirectToAction("PeekTempData");
+        }
+
+        public IActionResult PeekTempData()
+        {
+            var peekValue = TempData.Peek("key");
+            var getValue = TempData["key"];
```
Comment (rynowak) [inline test/WebSites/TempDataWebSite/Controllers/HomeController.cs]: Doesn't a peek followed by a get have different semantics than peek alone?

Comment (rynowak) [inline test/WebSites/TempDataWebSite/Controllers/HomeController.cs]: As discussed, we should have a test that doing a Peek keeps the value for the next request and doing a Get does not.

Comment (Eilon) [inline test/WebSites/TempDataWebSite/Controllers/HomeController.cs]: Indeed. The "key" test is this:

1. First request comes in, sets some TempData, does a redirect (or whatever)
2. Second request gets the TempData, confirms it is valid
3. Third request tries to get the TempData again, ensures that there *isn't* any TempData

That's the tricky thing with TempData: it takes three requests to verify the main scenario.

File: src/Microsoft.AspNet.Mvc.Core/SessionStateTempDataProvider.cs [hunk 2]
```diff
@@ -43,6 +43,11 @@ public class SessionStateTempDataProvider : ITempDataProvider
                 // If we got it from Session, remove it so that no other request gets it
                 session.Remove(TempDataSessionStateKey);
             }
+            else
+            {
+                // Initializing empty session so that it is established before the headers are sent.
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/SessionStateTempDataProvider.cs]: Add to this comment - we're calling `Save()` after the response has been sent which is why need the placeholder.

**Comment by rynowak** [comment]
:shipit:

**Pull request merged**